src/gui/notifications/notifier.cpp

```diff
@@ -0,0 +1,97 @@
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * In addition, as a special exception, the copyright holders give permission to
+ * link this program with the OpenSSL project's "OpenSSL" library (or with
+ * modified versions of it that use the same license as the "OpenSSL" library),
+ * and distribute the linked executables. You must obey the GNU General Public
+ * License in all respects for all of the code used other than "OpenSSL".  If you
+ * modify file(s), you may extend this exception to your version of the file(s),
+ * but you are not obligated to do so. If you do not wish to do so, delete this
+ * exception statement from your version.
+ */
+
+#include "notifier.h"
+
+#include "core/bittorrent/torrenthandle.h"
+#if (defined(Q_OS_UNIX) && !defined(Q_OS_MAC)) && defined(QT_DBUS_LIB)
+#include "dbusnotifier.h"
+#else
+#include "systemtraynotifier.h"
+#endif
+
+#include <QWidget>
+
+Notifier::Notifier(QObject *parent)
+    : QObject(parent)
+{
+}
+
+namespace
+{
+    inline QString torrentNameOrDefault(const BitTorrent::TorrentHandle* h)
+    {
+        return h ? h->name() : QObject::tr("Torrent");
+    }
+}
+
+void Notifier::notifyTorrentFinished(const BitTorrent::TorrentHandle *h, QWidget *widget)
+{
+    Context ct(widget, h);
+    showNotification(Notifier::TorrentFinished, tr("Download completion"),
+                     tr("%1 has finished downloading.", "e.g: xxx.avi has finished downloading.")
+                        .arg(torrentNameOrDefault(h)), &ct);
+}
+
+void Notifier::notifyTorrentIOError(const BitTorrent::TorrentHandle *h, const QString &message, QWidget *widget)
+{
+    Context ct(widget);
+    showNotification(Notifier::TorrentIOError, tr("I/O Error", "i.e: Input/Output Error"),
+                     tr("An I/O error occurred for torrent %1.\n Reason: %2",
+                        "e.g: An error occurred for torrent xxx.avi.\n Reason: disk is full.")
+                        .arg(torrentNameOrDefault(h)).arg(message), &ct);
+}
+
+void Notifier::notifyUrlDownloadError(const QString &url, const QString &reason, QWidget *widget)
+{
+    Context ct(widget);
+    showNotification(Notifier::UrlDownloadError, tr("Url download error"),
+                     tr("Couldn't download file at url: %1, reason: %2.").arg(url).arg(reason), &ct);
+}
+
+void Notifier::notifySearchFinished(QWidget *widget)
+{
+    Context ct(widget);
+    showNotification(Notifier::SearchFinished, tr("Search Engine"), tr("Search has finished"), &ct);
+}
+
+void Notifier::showNotification(const QString &title, const QString &message, QWidget *widget)
+{
+    Context ct(widget);
+    showNotification(Notifier::NoType, title, message, &ct);
+}
+
+Notifier *createNotifier(QWidget *mainWindow, QSystemTrayIcon *systrayIcon)
```

Comment:
GUI-less version will still require notifier creation in another place.
GUI-less version will still require more complex code rewriting so you shouldn't worry about it now.

And yet... I'm more inclined to believe the conclusion that it would be better to use instances of different classes in different qBittorrent build variants. Perhaps you should leave a common base class for them with some (really) common functionality in it (but only to prevent duplication of code), but use these objects through interfaces of subclasses (although it would require the use of ifdefs in the code).

>When submitting this PR I decided to not stress the intention to make it work in no-gui mode, but to submit a code, suitable for such modification. Instead, we can work on that right now and merge more changes at once.

I don't think this PR for immediate merge. So you should think about non-GUI code. Perhaps this will put some things in their places.